src/block_cache/block_mgr.c

```diff
@@ -182,6 +182,69 @@ __bm_checkpoint_unload(WT_BM *bm, WT_SESSION_IMPL *session)
     return (ret);
 }
 
+/*
+ * __bm_close_block_remove --
+ *     Remove a single block handle.
+ */
+static int
+__bm_close_block_remove(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    u_int i;
+
+    /* Discard any references we're holding. */
+    for (i = 0; i < block->related_next; ++i) {
+        --block->related[i]->ref;
+        block->related[i] = NULL;
+    }
+
+    /* Discard the block structure. */
+    return (__wt_block_close(session, block));
+}
+
+/*
+ * __bm_close_block --
+ *     Close a single block handle, removing the handle if it's no longer useful.
+ */
+static int
+__bm_close_block(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    bool found;
+
+    __wt_verbose(session, WT_VERB_BLKCACHE, "close: %s", block->name);
+
+    conn = S2C(session);
+
+    /* You can't close files during a checkpoint. */
+    WT_ASSERT(
+      session, block->ckpt_state == WT_CKPT_NONE || block->ckpt_state == WT_CKPT_PANIC_ON_FAILURE);
```

Comment:
Do we need to take a checkpoint lock here to really prevent this?

Reply:
I don't think so -- I put this in as a debugging hack because I got the reference counting wrong, and figured I'd leave it to guard against future missteps.